[Windows,Linux] [Input] Return the keyboard ID for raw input

knn217 opened the pull request:
Based on [#13083](https://github.com/godotengine/godot-proposals/issues/13083)

- Return the keyboard ID from raw input in InputEventKey as `device`
- When adding Input mapping, the Input events's `device` will be normalized to `ALL_DEVICE`, 
- Normalizing to a default `device` value for input map id safer, since raw input's id can change in runtime by unplugging/plugging back the keyboard
- If users wish to separate by keyboard ID, they need to get the event 's `device` to tell which keyboard the event is from, and handle hot-remapping if the ID was changed in runtime
- Added in 2nd commit: The returned `device` is now ordered starting from 0, and registered by plugin/input. Unplugging a keyboard will leave its `device` open for a new keyboard to register to

Test with GDScript:
```
extends Node

func _input(event):
	if event is InputEventKey and event.pressed and not event.echo:
		print("Pressed: ", OS.get_keycode_string(event.keycode), "| ID: ", event.device)
```

<img width="1127" height="586" alt="image" src="https://github.com/user-attachments/assets/db97525b-9dfd-4bf1-80c8-9f20be083440" />



Test with GDExtension:
```
void App::_input(const godot::Ref<godot::InputEvent> &event)
{
    godot::Ref<godot::InputEventKey> key_event = event;
    if (key_event.is_valid()){
        if (key_event->is_pressed() && !key_event->is_echo()) {
            godot::String key_name = godot::OS::get_singleton()->get_keycode_string(key_event->get_keycode());
            std::int64_t keyboard_id = key_event->get_device();
            godot::UtilityFunctions::print("Pressed: ", key_name, " | ID: ", keyboard_id);
        }
    }
}
```
<img width="1000" height="480" alt="image" src="https://github.com/user-attachments/assets/01390f91-2847-41ac-95e8-41c2849481cf" />

<img width="670" height="667" alt="image" src="https://github.com/user-attachments/assets/dc534c77-4323-4d4c-97b4-84dc677336a5" />


Test input mapping:
Tested action created from GDScript, with an event for pressing key right, the action triggered with multiple keyboard ids (0, 1)
```
extends Node2D

func _ready():
	InputMap.add_action("ui_right_custom")
	var event := InputEventKey.new()
	event.physical_keycode = KEY_RIGHT
	event.pressed = true
	event.echo = false
	InputMap.action_add_event("ui_right_custom", event)

func _process(_delta):
	if Input.is_action_pressed("ui_right_custom", false):
		print("GDScript moving right")

func _input(event):
	if event is InputEventKey:
		if not event.echo:
			if event.pressed:
				print("GDScript Pressed: ", OS.get_keycode_string(event.keycode), "| ID: ", event.device)
			else:
				print("GDScript Released: ", OS.get_keycode_string(event.keycode), "| ID: ", event.device)

```
<img width="1228" height="887" alt="image" src="https://github.com/user-attachments/assets/f90496bf-516a-46b1-b123-f8f38bc78f8a" />

Tested with an action from the input map setting
```
extends Node2D

func _process(_delta):
	if Input.is_action_pressed("ui_left_custom", false):
		print("GDScript moving left")

func _input(event):
	if event is InputEventKey:
		if not event.echo:
			if event.pressed:
				print("GDScript Pressed: ", OS.get_keycode_string(event.keycode), "| ID: ", event.device)
			else:
				print("GDScript Released: ", OS.get_keycode_string(event.keycode), "| ID: ", event.device)

```
<img width="913" height="455" alt="image" src="https://github.com/user-attachments/assets/eeb006db-44c3-443b-9fe9-e476b0638067" />

<img width="1247" height="773" alt="image" src="https://github.com/user-attachments/assets/b35a0ca3-4cd9-40da-9e62-258a5972e2b4" />

Test on Linux X11 (Ubuntu 24.04):
<img width="1117" height="793" alt="image" src="https://github.com/user-attachments/assets/6648acc1-65a5-458d-b7b1-543d2ba2ddca" />


- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/13083.*

bruvzg commented on core/input/input_event.cpp:
```diff
@@ -360,6 +360,15 @@ Key InputEventKey::get_physical_keycode() const {
 	return physical_keycode;
 }
 
+void InputEventKey::set_keyboard_id(int64_t p_id) {
```
It would make more sense to have `device_id` (or use it with existing `device`, but this will require more changes) in the base `InputEvent`, the same can be used for mouse as well.

For the reference, native device ID on macOS can be obtained from keyboard/mouse events, using the following:
```obj-c++
int64 _get_device_id(NSEvent *p_event) {
	return CGEventGetIntegerValueField([p_event CGEvent], /*kCGEventRegistryID*/ CGEventField(87));
}
```

knn217 replied:
Ah, that makes sense.
Should've look deeper into the parents.
Do I need to wait until check flows are completed to push?

bruvzg replied:
> Do I need to wait until check flows are completed to push?

No, it will cancel currently running tasks if you push.

bruvzg commented:
Simply using `set_device` will likely break a lot of stuff, input map actions default to device `0` when matching actions, so won't work with any other ID, and probably should be changed to default to `-1` (all devices, which might conflict with `DEVICE_ID_EMULATION` with is also -1).

knn217 commented:
I looked into InputMap and it seems safer to only use ALL_DEVICE for input mapping, since keyboard ID can change when user unplug -> plug keyboard back in, which invalidates the mapping if we use keyboard ID (device)
For now, I normalize InputEventKey when it is added/searched in input map

> probably should be changed to default to -1 (all devices, which might conflict with DEVICE_ID_EMULATION with is also -1).

I made the change and did not find any issue.

Gnumaru commented:
@knn217 In your first screenshot there are three diferent keyboard ids. Is this because you really plugged 3 different keyboards to test, or perhaps windows simply did changed the keyboard id in between keystrokes?

Also, does the windows api expose some way of knowing the device ordering? to know what is the first, the second, the third keyboard and os on? the device ids for joypads that godot returns are always ordered starting from zero, so that if you have 4 joypads they will have device ids 0, 1, 2 and 3. Is it possible to at least emulate this with keyboards? even if the OSs api don't tell you the keyboard ordering, couldn't we store a dictionary mapping the raw keyboard ids returned by the OS into a zero based ordered id? even if we just make up this ordering ourselves? for example, if windows returns ids 687, -94601 and 12687 whe order them ascendingly and give them the ids 0 (for -94601), 1 (for 687) and 2 (for 12687).

knn217 commented:
@Gnumaru 
> @knn217 In your first screenshot there are three diferent keyboard ids. Is this because you really plugged 3 different keyboards to test, or perhaps windows simply did changed the keyboard id in between keystrokes?

Windows can change the keyboard ID (handle), but only when user unplug -> plug keyboard back in. The ID stays the same while plugged in though.
I have updated the test to show this, by unpluging the keyboard, I was able to change the keyboard ID 10 times

> Also, does the windows api expose some way of knowing the device ordering? to know what is the first, the second, the third keyboard and os on? the device ids for joypads that godot returns are always ordered starting from zero, so that if you have 4 joypads they will have device ids 0, 1, 2 and 3

I have not found any API like that.

> Is it possible to at least emulate this with keyboards? even if the OSs api don't tell you the keyboard ordering, couldn't we store a dictionary mapping the raw keyboard ids returned by the OS into a zero based ordered id?
Like I have shown before, after unplugged, Windows assigns a new ID to the keyboard when plugged back in, making the previous ID deprecated.
I think it's possible to emulate this behavior, by using this API: 
https://learn.microsoft.com/en-us/windows/win32/inputdev/wm-input-device-change
I have not used this myself but this seems to provide notifications for whenever a device is plugged/unplugged

Here's how I understand the flowchart should be implemented from your description:
- device plugged in -> Add to map { "raw kb id": "lowest available slot" } ->update "lowest available slot"
- device unplugged -> Remove from map at key "raw kb id" ->update "lowest available slot"
- If a key is pressed, contains an ID not existed in map -> insert it the same as "device plugged in"

The last step is to solve the issue with laptop's keyboards being always plugged in, and also any keyboards that are already plugged in.

knn217 commented:
I have added another commit to support:

- Adding the same functionality on Linux (only X11, Wayland doesn't have support for this as far as I know)
- Register ID is now only by pressing a key on a keyboard, NOT plugging in.
- I removed registering by plugging since on Linux since a lot of devices showed up as keyboard. I couldn't find a way to differentiate the actual physical keyboard from the members in `XIDeviceInfo`
- I also removed registering by plugging on Windows to sync the functionality with Linux. Now on register by key press is supported

Here's the test on Linux X11 (Ubuntu 24.04):
<img width="1117" height="793" alt="image" src="https://github.com/user-attachments/assets/0d3ddb88-a885-445c-86fc-6cc4f4501fd5" />

AThousandShips commented on servers/display/display_server.cpp:
```diff
@@ -491,6 +491,31 @@ bool DisplayServer::_get_window_early_clear_override(Color &r_color) {
+	return;
+}
+
+void DisplayServer::kb_map_remove(int raw_kb_id) {
```
```suggestion
void DisplayServer::kb_map_remove(int p_raw_kb_id) {
```

AThousandShips commented on servers/display/display_server.cpp:
```diff
@@ -491,6 +491,31 @@ bool DisplayServer::_get_window_early_clear_override(Color &r_color) {
+			break;
+		}
+	}
+	return;
```
```suggestion
```

AThousandShips commented on servers/display/display_server.cpp:
```diff
@@ -491,6 +491,31 @@ bool DisplayServer::_get_window_early_clear_override(Color &r_color) {
+	}
+	// Mark as unused to delete later
+	id_map[raw_kb_id] = false;
+	return;
```
```suggestion
```

AThousandShips commented on platform/linuxbsd/x11/display_server_x11.cpp:
```diff
@@ -4840,6 +4867,12 @@ void DisplayServerX11::process_events() {
+					case XI_RawKeyPress:
+					case XI_RawKeyRelease: {
+						XIRawEvent *raw_event = (XIRawEvent *)event_data;
+						this->kb_map_insert(raw_event->sourceid);
```
```suggestion
						kb_map_insert(raw_event->sourceid);
```

AThousandShips commented on platform/linuxbsd/x11/display_server_x11.cpp:
```diff
@@ -251,6 +251,26 @@ bool DisplayServerX11::_refresh_device_info() {
+		}
+		// Only remove the 1st ID that is not listed
+		if (!listed) {
+			this->kb_map_remove(raw_kb_id);
```
```suggestion
			kb_map_remove(raw_kb_id);
```

knn217 commented:
> Hi, Wayland should support this behavior already. Compositors can expose multiple `wl_seat` globals, each with a different keyboard, mouse, and touchscreen. Note though that some compositors (e.g. sway) merge multiple devices under a single seat unless the user says otherwise. This is not mandatory according to the spec, it's just compositor-specific policy.

I see, but is there a way to make the compositor expose seats from Godot? Compositors are outside of application's control as I understand.

deralmas commented:
> I see, but is there a way to make the compositor expose seats from Godot? Compositors are outside of application's control as I understand.

Hi, I'm not really sure what you mean. Godot can't play with input settings as they're non-standard and outside of the scope of the Wayland protocol. I also don't know how other compositors outside of sway handle it, they could as well make a new seat every time, it's up to them.

If the compositor is configured (by default or by the user) to expose multiple input devices, godot will be able to detect and use all of them. In fact, we already do that, we already have logic for handling multiple input devices as we're required to do that.

knn217 commented:
Support for Wayland seems too complicated for me currently, so I will leave the PR here for now.
Maybe Wayland support can be added in another PR.

> If the compositor is configured (by default or by the user) to expose multiple input devices, godot will be able to detect and use all of them. In fact, we already do that, we already have logic for handling multiple input devices as we're required to do that.


@deralmas 
I'm still new to Godot so I'm not sure what this logic is for. Please let me know what feature this is and how to set it up. I would like to play around with this first to see how Godot handle it and what compositor config Godot is expecting.
Thank you!

knn217 committed:
[Input] [Windows] [X11] [Wayland] Support device indexing for inputs from multiple keyboards (seats in Wayland case)
